smithy-linters/src/main/java/software/amazon/smithy/linters/DeprecatedTraitsValidator.java

```diff
@@ -0,0 +1,72 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.linters;
+
+import static java.lang.String.format;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.traits.DeprecatedTrait;
+import software.amazon.smithy.model.traits.TraitDefinition;
+import software.amazon.smithy.model.validation.AbstractValidator;
+import software.amazon.smithy.model.validation.ValidationEvent;
+import software.amazon.smithy.model.validation.ValidatorService;
+
+/**
+ * Emits a validation event if a model contains shapes that are bound to deprecated traits.
+ */
+public final class DeprecatedTraitsValidator extends AbstractValidator {
+
+    public static final class Provider extends ValidatorService.Provider {
+        public Provider() {
+            super(DeprecatedTraitsValidator.class, DeprecatedTraitsValidator::new);
+        }
+    }
+
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        Set<ShapeId> deprecatedTraits = model.shapes()
+                .filter(shape -> shape.hasTrait(TraitDefinition.class))
+                .filter(trait -> trait.hasTrait(DeprecatedTrait.class))
+                .map(Shape::getId)
+                .collect(Collectors.toSet());
+
+        return model.shapes()
+                .flatMap(shape -> validateShape(model, shape, deprecatedTraits).stream())
+                .collect(Collectors.toList());
+    }
+
+    private List<ValidationEvent> validateShape(Model model, Shape shape, Set<ShapeId> deprecatedTraits) {
+        List<ValidationEvent> events = new ArrayList<>();
+        shape.getAllTraits().forEach((shapeId, trait) -> {
+            if (deprecatedTraits.contains(trait.toShapeId())) {
+                DeprecatedTrait deprecatedTrait = model.expectShape(trait.toShapeId()).getTrait(DeprecatedTrait.class)
+                        .get();
+                String traitMessage = trait.toShapeId().toString();
+                if (deprecatedTrait.getMessage().isPresent()) {
+                    traitMessage = traitMessage.concat(", " + deprecatedTrait.getMessage().get());
+                }
+                events.add(warning(shape, format("This shape applies a trait that is deprecated: %s", traitMessage)));
```

Comment:
This should be relative to the trait that was applied to the shape. You can do this by passing then trait in as the second argument.

```events.add(warning(shape, trait, format(...```